crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -0,0 +1,748 @@
+    case W4.asNat (C.llvmPointerBlock val) of
+      Just 0 -> return $ MapF.Pair join_varaible $ C.llvmPointerOffset val
+      _ -> fail $ "SimpleLoopFixpoint.loadMemJoinVariables: unexpected val:" ++ show (C.ppPtr val))
+  (Map.toAscList subst)
+
+storeMemJoinVariables ::
+  (C.IsSymInterface sym, C.HasPtrWidth wptr, C.HasLLVMAnn sym) =>
+  sym ->
+  C.MemImpl sym ->
+  Map (Natural, Natural, Natural) (MemFixpointEntry sym, C.StorageType) ->
+  MapF (W4.SymExpr sym) (W4.SymExpr sym) ->
+  IO (C.MemImpl sym)
+storeMemJoinVariables sym mem mem_subst eq_subst = foldlM
+  (\mem_acc ((blk, off, _sz), (MemFixpointEntry { memFixpointEntryJoinVariable = join_varaible }, storeage_type)) -> do
+    ptr <- C.LLVMPointer <$> W4.natLit sym blk <*> W4.bvLit sym ?ptrWidth (BV.mkBV ?ptrWidth $ fromIntegral off)
+    C.doStore sym mem_acc ptr (C.LLVMPointerRepr $ W4.bvWidth join_varaible) storeage_type C.noAlignment =<<
+      C.llvmPointer_bv sym (MapF.findWithDefault join_varaible join_varaible eq_subst))
+  mem
+  (Map.toAscList mem_subst)
+
+dropMemStackFrame :: C.IsSymInterface sym => C.MemImpl sym -> (C.MemImpl sym, C.MemAllocs sym, C.MemWrites sym)
+dropMemStackFrame mem = case (C.memImplHeap mem) ^. C.memState of
+  (C.StackFrame _ _ _ (a, w) s) -> ((mem { C.memImplHeap = (C.memImplHeap mem) & C.memState .~ s }), a, w)
+  _ -> C.panic "SimpleLoopFixpoint.dropMemStackFrame" ["not a stack frame:", show (C.ppMem $ C.memImplHeap mem)]
+
+
+filterSubstitution ::
+  C.IsSymInterface sym =>
+  sym ->
+  MapF (W4.SymExpr sym) (FixpointEntry sym) ->
+  MapF (W4.SymExpr sym) (FixpointEntry sym)
+filterSubstitution sym substitution =
+  -- TODO: fixpoint
+  let uninterp_constants = foldMapF
+        (Set.map (C.mapSome $ W4.varExpr sym) . W4.exprUninterpConstants sym . bodyValue)
+        substitution
+  in
+  MapF.filterWithKey (\variable _entry -> Set.member (C.Some variable) uninterp_constants) substitution
+
+uninterpretedConstantEqualitySubstitution ::
+  forall sym .
+  C.IsSymInterface sym =>
+  sym ->
+  MapF (W4.SymExpr sym) (FixpointEntry sym) ->
+  (MapF (W4.SymExpr sym) (FixpointEntry sym), MapF (W4.SymExpr sym) (W4.SymExpr sym))
+uninterpretedConstantEqualitySubstitution _sym substitution =
+  let reverse_substitution = MapF.foldlWithKey'
+        (\accumulator variable entry -> MapF.insert entry variable accumulator)
+        MapF.empty
+        substitution
+      uninterpreted_constant_substitution = fmapF
+        (\entry -> fromJust $ MapF.lookup entry reverse_substitution)
+        substitution
+      normal_substitution = MapF.filterWithKey
+        (\variable _entry ->
+          Just Refl == W4.testEquality variable (fromJust $ MapF.lookup variable uninterpreted_constant_substitution))
+        substitution
+  in
+  (normal_substitution, uninterpreted_constant_substitution)
+
+
+userSymbol' :: String -> W4.SolverSymbol
+userSymbol' = fromRight (C.panic "SimpleLoopFixpoint.userSymbol'" []) . W4.userSymbol
+
+
+simpleLoopFixpoint ::
```

Comment:
It would be really useful to have a description of the intent of this code and an indication as to how a user should use it.